databricks-sdk-java/src/main/java/com/databricks/sdk/core/NotebookNativeCredentialsProvider.java

```diff
@@ -0,0 +1,133 @@
+package com.databricks.sdk.core;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A CredentialsProvider that uses the API token from the command context to authenticate.
+ *
+ * <p>The token and hostname are read from the command context, which can be retrieved through the
+ * dbutils API. As the Java SDK does not depend on DBUtils directly, reflection is used to retrieve
+ * the token. This token should be available wherever the DBUtils API is accessible (i.e. in the
+ * Spark driver).
+ */
+public class NotebookNativeCredentialsProvider implements CredentialsProvider {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(NotebookNativeCredentialsProvider.class);
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+
+  @Override
+  public String authType() {
+    return "runtime";
+  }
+
+  @Override
+  public HeaderFactory configure(DatabricksConfig config) {
+    if (System.getenv("DATABRICKS_RUNTIME_VERSION") == null) {
+      return null;
+    }
+
+    // DBUtils is not available in the Java SDK, so we have to use reflection to get the token.
+    // First, we get the context by calling getContext on the notebook field of dbutils, then we get
+    // the apiKey and
+    // apiUrl fields from the context. If this is successful, we set the host on the config.
+    try {
+      Object dbutils = getDbUtils();
+      Object notebook = getField(dbutils, "notebook");
+      TokenAndUrl testTokenAndUrl = getTokenAndUrl(notebook);
+      if (testTokenAndUrl.url == null) {
+        LOG.debug("Workspace URL is not available, skipping runtime auth");
+      }
+      config.setHost(testTokenAndUrl.url);
+
+      return () -> {
+        Map<String, String> headers = new HashMap<>();
+        TokenAndUrl tokenAndUrl;
+        try {
+          tokenAndUrl = getTokenAndUrl(notebook);
+        } catch (IllegalAccessException
+            | NoSuchMethodException
+            | InvocationTargetException
+            | JsonProcessingException e) {
+          throw new RuntimeException(e);
+        }
+        headers.put("Authorization", String.format("Bearer %s", tokenAndUrl.token));
+        return headers;
+      };
+    } catch (Exception e) {
+      LOG.debug("Failed to get token from dbutils", e);
+      return null;
+    }
+  }
+
+  /** Load the dbutils object initialized by DBR. */
+  private static Object getDbUtils() {
+    try {
+      Class<?> dbutilsHolderClass = Class.forName("com.databricks.dbutils_v1.DBUtilsHolder$");
+      Object dbutilsHolder = dbutilsHolderClass.getDeclaredField("MODULE$").get(null);
+      InheritableThreadLocal<Object> dbutils = getField(dbutilsHolder, "dbutils0");
+      return dbutils.get();
+    } catch (Exception e) {
```

Comment:
Catching a specific exception might be better

Reply:
Can you give a concrete example? The issue is that we actually do want to catch all exceptions, and we don't want to handle any exceptions differently. The codepath should always be "rethrow as DatabricksException", and then `configure()` can have a top-level path of just "log at debug and return null".

Reply:
I need to check for the underlying methods for concrete exceptions we should catch but I think catching all exception and throwing them as DatabricksException isn't useful for example -- if there is NullPointerException (which is a RuntimeException) but we throw that wrapped in DatabricksException. It would pollute the stack trace and in most cases the NullPointerException happens because of a bug -- catching this would hide (to some extent) the underlying bug. 

> The codepath should always be "rethrow as DatabricksException", and then configure() can have a top-level path of just "log at debug and return null".
But we should do this only for specific exceptions right and not for all exceptions?

Reply:
In the auth codepath, the contract is that `configure()` returns null to indicate that the credential provider can't be used with the given `DatabricksConfig`. Throwing an exception aborts the `configure()` method. Because of this, we explicitly do not want to throw exceptions across this error boundary. Instead, we can log exceptions as they are thrown at DEBUG so that users can see them if needed for debugging. Note that I'm never discarding the original exception; it is listed as the cause of the DatabricksException, so it will be visible in the `Caused by` section of the stack trace when printed out.

Reply:
What specific exception would we want to handle differently, and how should we handle it?

Reply:
That's fair. We would still catch exceptions broadly at the root. I can make sure to catch the checked exceptions and do some null checking. I just need to test this carefully to make sure this doesn't cause issues with stability (though, I guess it is the last provider in DefaultCredentialsProvider, so even if it fails, it won't block other providers from loading).